### PR TITLE
Provide possibility to extend sure. Refs #31

### PR DESCRIPTION
**Pull request by timofurrer**
*__Note__: work in progress*

This PR contains the following features:

- [x] Add decorator to extend sure with a custom assertion method
- [x] Add decorator to extend sure with a chain method
- [ ] Add decorator to extend sure with a chain property

**Other Todos:**

- [ ] Enforce beauty of assertion error message. How? Provide specific `Exception`: `class Assertion(AssertionError): pass` which takes *expected* and *given* arguments? Is it just best practice?
- [ ] Enforce negative check. How? Is it just best practice?
- [ ] Documentation / Tutorial

## Custom assertion example

```python
from sure import assertion


class Response(object):
    def __init__(self, return_code):
        self.return_code = return_code


@assertion
def return_code(self, return_code):
    if self.negative:
        msg = "Expected was a return code different from {0}."
        assert return_code != self.obj.return_code, msg.format(return_code)
    else:
        msg = "Expected return code is: {0}\nGiven return code was: {1}"
        assert return_code == self.obj.return_code, msg.format(return_code, 
            self.obj.return_code)

Response(200).should.have.return_code(200)
Response(200).shouldnt.have.return_code(201)
```

@gabrielfalcao please provide feedback regarding API. See first test for entire example usage.

Refs: #31

### CODE REVIEW

**Comment by timofurrer** [comment]
@gabrielfalcao Would you prefer to pass important values to the `assertion function` directly?

The interface would then look like:

```python
@assertion
def return_code(self, what, is_negative, return_code):
    pass
```

This would have the benefit that the user does not need to know the `AssertionBuilder` class that well. The user would get an impression what's important for the assertion from the methods interface.

However, with that the user is kinda limited to those values - even worse would be if the user would not need the passed values but instead needs to access something from `self`: every linter would complain about unused arguments.

File: sure/__init__.py [hunk 1]
```diff
@@ -925,6 +925,13 @@ def match(self, regex, *args):
 expect = AssertionBuilder('expect')
 
 
+def assertion(func):
+    """Extend sure with a custom assertion method."""
+    func = assertionmethod(func)
+    setattr(AssertionBuilder, func.__name__, func)
```
Comment (timofurrer) [inline sure/__init__.py]: Maybe don't use `setattr` but registration of assertion into a dict and then use `__getattr__` ...?

File: tests/test_custom_assertions.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+            self.return_code = return_code
+
+
+    @chain
+    def header(self, header_name):
+        self.obj.headers.should.have.key(header_name)
+        return self.obj.headers[header_name]
+
+
+    # FIXME(TF): 'must' does not sound right in this method chain.
```
Comment (timofurrer) [inline tests/test_custom_assertions.py]: @gabrielfalcao what do you think about adding a `which` or `having` to the list of `POSITIVES` for the monkey patched PyObject properties?

**Comment by gabrielfalcao** [comment]
This developer experience is fantastic.

**Pull request merged**